src/hardware/sblaster.cpp

```diff
@@ -1013,19 +1013,65 @@ static std::vector<AudioFrame>& maybe_silence(const T* samples,
 
 static uint32_t read_dma_8bit(const uint32_t bytes_to_read, const uint32_t i = 0)
 {
-	const auto bytes_read = sb.dma.chan->Read(bytes_to_read, sb.dma.buf.b8 + i);
-	assert(bytes_read <= DmaBufSize * sizeof(sb.dma.buf.b8[0]));
+	static_assert(sizeof(sb.dma.buf.b8[0]) == 1);
+
+	if (i >= DmaBufSize) {
+		// Should never happen as the code is current written.
+		// Calling code has i either 0 or 1 to handle a dangling sample from the last read.
+		assertm(false, "SBLASTER: Read requested out of bounds of sb.dma.buf.b8");
+		return 0;
+	}
+
+	// Assert that the DMA controller is configured for 8 bit reads.
+	// so we don't have to do the whole song and dance we do for 16 bit.
+	// We should always be dealing with bytes here.
+	assertm(!sb.dma.chan->is_16bit, "SBLASTER: read_dma_8bit() called but DMA controller is configured for 8 bit reads");
+
+	const uint32_t bytes_available = DmaBufSize - i;
+	const uint32_t clamped_bytes = std::min(bytes_to_read, bytes_available);
+	const auto bytes_read = sb.dma.chan->Read(clamped_bytes, sb.dma.buf.b8 + i);
+	assert(bytes_read <= clamped_bytes);
 
 	return check_cast<uint32_t>(bytes_read);
 }
 
-static uint32_t read_dma_16bit(const uint32_t bytes_to_read, const uint32_t i = 0)
+static uint32_t read_dma_16bit(const uint32_t words_to_read, const uint32_t i = 0)
 {
+	static_assert(sizeof(sb.dma.buf.b16[0]) == 2);
+
+	if (i >= DmaBufSize) {
+		// Should never happen as the code is current written.
+		// Calling code has i either 0 or 1 to handle a dangling sample from the last read.
+		assertm(false, "SBLASTER: Read requested out of bounds of sb.dma.buf.b16");
+		return 0;
+	}
+
+	// In dma.cpp, if sb.dma.chan->is_16bit is set, we're dealing with 16-bit words.
+	// Otherwise, we're dealing with 8 bit words (bytes).
+	// Calling code handles this case and conditionally divides by two.
+	uint32_t bytes_requested = words_to_read;
+	if (sb.dma.chan->is_16bit) {
+		bytes_requested *= 2;
+		// Calling code uses these internal SoundBlaster variables.
+		// sb.dma.chan->is_16bit is the source of truth in terms of memory safety.
+		// Assert that these two always match up.
+		assertm(sb.dma.mode == DmaMode::Pcm16Bit, "SoundBlaster expected 8 bit read but DMA controller is 16 bit");
+	} else {
+		assertm(sb.dma.mode == DmaMode::Pcm16BitAliased, "SoundBlaster expected 16 bit read but DMA controller is 8 bit");
+	}
+
+	// Clamp words to read so we don't overflow our buffer.
+	const uint32_t bytes_available = (DmaBufSize - i) * 2;
+	uint32_t clamped_words = std::min(bytes_requested, bytes_available);
+	if (sb.dma.chan->is_16bit) {
+		clamped_words /= 2;
```

Comment:
Assertion opportunity: check that `clamped_words` is even?

Reply:
Let me re-visit this.  It's not a problem if there's an odd number of bytes available in the buffer as long as we're doing mono sound.  We just round down the number of 16-bit words to read.

I think the calling code has a bug for stereo though where it could play silence in one channel for a frame if this happens.  And then also potentially invert the channels the next time it runs due to this weird "dangling sample" edge case.

Reply:
> The "dangling sample" code handles exactly this case. I verified it always enqueues whole frames into the mixer. If DMA returns an odd number of samples, it moves that odd sample to the front of the buffer so it gets used on the next call.

Is there a comment explaining this where the dangling sample is added? If not, could you add a short version of the above explanatory comment there? It would be super useful for the next guy doing a deep dive into the mixer code 🤿 🌊 

>  This seems unlikely to ever happen to me but who knows there's probably a weird demo out there that does it 😆

I'd bet good money on that! 😂

Reply:
Yeah there's a comment already.  I just didn't see it at first:  https://github.com/dosbox-staging/dosbox-staging/blob/33a83fb77173a590b6a89be35bc22cc6bb867806/src/hardware/sblaster.cpp#L1120-L1138